### PR TITLE
Implementation of poke API client using go

### PR DESCRIPTION
**Pull request by abdahmed22**
https://github.com/codescalersinternships/home/issues/224

### CODE REVIEW

File: .github/workflows/fmt.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    steps:
+    - uses: actions/checkout@v4
+    
+    - name: Install Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: ${{ env.GO_VERSION }}
+        
+    - name: Checkout code
+      uses: actions/checkout@v4
```
Comment (Omarabdul3ziz) [inline .github/workflows/fmt.yml]: why check out again?

File: .github/workflows/fmt.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      
+    - name: Install dependencies
+      run: make install_dependencies
```
Comment (Omarabdul3ziz) [inline .github/workflows/fmt.yml]: is it necessary to install deps? it is just a format

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    steps:
+    - uses: actions/checkout@v4
+    
+    - name: Install Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: ${{ env.GO_VERSION }}
+        
+    - name: Checkout code
+      uses: actions/checkout@v4
```
Comment (Omarabdul3ziz) [inline .github/workflows/lint.yml]: checkout at the start of the job is enough

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,19 @@
+all: install_dependencies build_binaries format lint test 
+
+install_dependencies:
+	go get ./...
```
Comment (Omarabdul3ziz) [inline Makefile]: `go mod tidy` helps with cleaning the mod files too, please check the difference

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+type BodyFormat struct {
+	Pokemons []Pokemon `json:"results"`
+}
+
+// NewClient creates a new client
+func NewClient(options ...Option) *Client {
+
+	err := godotenv.Load("../.env")
```
Comment (Omarabdul3ziz) [inline client/client.go]: client shouldn't handle reading envvars it should be read in the main and passed to the factory method

File: client/client.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		log.Fatalf("Error loading .env file: %s", err)
```
Comment (Omarabdul3ziz) [inline client/client.go]: don't panic in a pkg, you either handle it or return to main

File: client/poke.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/cenkalti/backoff"
+)
+
+// GetPokemonByName mimics a user performing a request to get a certain pokemon from the server
+func (c *Client) GetPokemonByName(ctx context.Context, pokemonName string) (Pokemon, error) {
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.URL+c.endPoint+"/"+pokemonName, nil)
+
+	if err != nil {
+		return Pokemon{}, err
```
Comment (Omarabdul3ziz) [inline client/poke.go]: it is good to add more context to your error, `fmt.Errorf("failed to create request: %w", err)`

File: client/poke.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		return resErr
+	}
+
+	notify := func(err error, t time.Duration) {
+		log.Printf("error: %v happened at time: %v", err, t)
+	}
+
+	err = backoff.RetryNotify(retryable, b, notify)
+	if err != nil {
+		log.Fatalf("error after retrying: %v", err)
```
Comment (Omarabdul3ziz) [inline client/poke.go]: don't panic in a package. return to main

File: client/poke.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+// GetAllPokemons mimics a user performing a request to get all pokemons from the server
+func (c *Client) GetAllPokemons(ctx context.Context) ([]Pokemon, error) {
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.URL+c.endPoint, nil)
+
+	if err != nil {
+		return []Pokemon{}, err
+	}
+
+	b := backoff.NewExponentialBackOff()
```
Comment (Omarabdul3ziz) [inline client/poke.go:77]: instead of duplicating the retrying code, you can have a helper method

File: main.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+		fmt.Println(returnedPokemon)
+	}
+
+	fmt.Println("GET /pokemon  -> return all pokemons infotmation")
+	returnedPokemons, err := myClient.GetAllPokemons(context.Background())
+
+	if err != nil {
+		log.Fatalf("error getting all pokemons: %s", err)
+	} else {
+		fmt.Println(returnedPokemons)
```
Comment (Omarabdul3ziz) [inline main.go:57]: avoid nesting as much as possible, no need to print the result in `else` block. 

```go
if err != nil {
  // handle err
}
// print result & continue
```
is the same

**Review by Omarabdul3ziz — CHANGES_REQUESTED** [review]

File: .github/workflows/fmt.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    - name: Run format
+      run: make format
+
+    - name: Check format
```
Comment (Omarabdul3ziz) [inline .github/workflows/fmt.yml:32]: it also run format right?

File: .github/workflows/lint.yml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+      
+    - name: Install dependencies
+      run: make install_dependencies
+
+
+    - name: Run linter
```
Comment (Omarabdul3ziz) [inline .github/workflows/lint.yml:33]: check `golangci-lint` action

File: main.go (hunk 10 above)
Comment (abdahmed22) [reply main.go:57]: I do not want to display the returned poke if error because it will be empty object

File: .github/workflows/fmt.yml (hunk 11 above)
Comment (abdahmed22) [reply .github/workflows/fmt.yml:32]: yes

File: .github/workflows/lint.yml (hunk 12 above)
Comment (abdahmed22) [reply .github/workflows/lint.yml:33]: I am using this inside the make file

**Pull request merged**